### PR TITLE
Defer flush of the policy resolver

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

Flushing after every match update causes high CPU usage for endpoints that are being added/removed with many matching policies.  Defer the sort and flush until we're ready to send data downstream.

## Todos

- [x] Tests
- [x] Documentation
- [x] Release note

## Release Note

```release-note
Fix slow performance in Felix calculation graph when adding/removing an endpoint with many matching policies. Policy sort order was rechecked once for each added/removed policy match rather than once only.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Comment by Josh-Tigera** [comment]
Was the inefficiency of the redundant flushes flagged by any tests? Would it be worthwhile either adding a benchmark test or an expectation of a canary-in-the-coal-mine function being called less than a threshold to protect against this kind of thing in the future?

**Comment by fasaxc** [comment]
I have a benchmark test in a follow-on PR.  It can't go in yet because it runs too slowly without the _other_ enhancements that I have lined up!

**Comment by fasaxc** [comment]
Added a basic kick-the-tires in the UT.

File: felix/calc/policy_resolver.go [hunk 1]
```diff
@@ -111,15 +111,13 @@ func (pr *PolicyResolver) OnUpdate(update api.Update) (filterOut bool) {
 			pr.markEndpointsMatchingPolicyDirty(key)
 		}
 	}
-	pr.maybeFlush()
```
Comment (aaaaaaaalex) [inline felix/calc/policy_resolver.go:114]: IIUC, the policyresolver now only flushes when the calcgraph does - are there any types of updates which should still be flushed more aggressively / decided explicitly by the pol-resolver / how much risk is there of the CG's flush strategy having a blind spot? More of a question than feedback :smile:

Comment (fasaxc) [reply felix/calc/policy_resolver.go:114]: It's also the case that the policy resolver is at the bottom of the calc graph already so it's the last thing in its call path that emits events (so there's no chance of reordering events here, I don't think)

**Review by Josh-Tigera — APPROVED** [review]

**Pull request merged**